### PR TITLE
add int8 qkv-fusion pass for x86

### PR DESCRIPTION
**Pull request by yuchengliu1**
This PR add qkv-fusion pass on x86 CPU. It added a manually registered qkv-usion pass. This fusion pass fuse 3 gemm(qkv) and SDPA to a gemm kernel and a SPDA kernel in int8.

The current fp8 mkl tensor does not support unpack. Thus, fp8 qkv-fusion will be add until pytorch updated.

### CODE REVIEW

**Review by Valentine233 — COMMENTED** [review]
Could you please add the UT?

**Comment by Valentine233** [comment]
Hi @yuchengliu1, any update?

File: test/prototype/inductor/test_qsdpa_fusion.py [hunk 1]
```diff
@@ -343,19 +370,22 @@ def _test_int8_sdpa_rewriter(self):
 
         # pattern is different for bs=1
         torch.manual_seed(1234)
-        for dtype, has_mask, bs in itertools.product(
-            [torch.float32, torch.bfloat16], [True, False], [56, 1]
-        ):
+        for dtype, has_mask, bs, fused in itertools.product(
+            [torch.float32], [True], [56], [True]
```
Comment (Valentine233) [inline test/prototype/inductor/test_qsdpa_fusion.py]: Why not `[torch.float32, torch.bfloat16], [True, False], [56, 1], [True, False]`?

Comment (yuchengliu1) [reply test/prototype/inductor/test_qsdpa_fusion.py]: fixed

**Review by Valentine233 — APPROVED** [review]

File: test/prototype/inductor/test_qsdpa_fusion.py [hunk 2]
```diff
@@ -245,6 +245,33 @@ def forward(self, x, mask):
         return self.dense(context_layer)
 
 
+class fuse_MHAModule(torch.nn.Module):
```
Comment (mingfeima) [inline test/prototype/inductor/test_qsdpa_fusion.py]:
```suggestion
class FusedMHAModule(torch.nn.Module):
```

can we rename it in this way?

File: test/prototype/inductor/test_qsdpa_fusion.py [hunk 3]
```diff
@@ -245,6 +245,33 @@ def forward(self, x, mask):
...
+    def __init__(
+        self,
+        input_dim,
+        has_mask,
+        num_attention_heads,
+        attention_head_size,
+    ) -> None:
+        super().__init__()
+        self.qkv_proj = torch.nn.Linear(input_dim * 3, input_dim * 3, bias=False)
```
Comment (mingfeima) [inline test/prototype/inductor/test_qsdpa_fusion.py]: I noticed that there is only `input_dim` in `MHAModule`, no `output_dim`.
I am not sure if this is the intended way, so we have 3 linears, each with [OC, IC] == [input_dim, input_dim]

for this one, qkv fusion indicates that OC * 3 but IC stays unchanged, [3OC, IC] == [3* input_dim, input_dim], no [3 * input_dim, 3 * input_dim]

**Review by mingfeima — CHANGES_REQUESTED** [review]

File: test/prototype/inductor/test_qsdpa_fusion.py (hunk 2 above)
Comment (yuchengliu1) [reply test/prototype/inductor/test_qsdpa_fusion.py]: I have renamed it

File: test/prototype/inductor/test_qsdpa_fusion.py (hunk 3 above)
Comment (yuchengliu1) [reply test/prototype/inductor/test_qsdpa_fusion.py]: I have adjust the IC of `qkv_proj`. `FusedMHA` has the same input as `MHA` now.
`input_dim` and `output_dim` of qkv is not the key point of MHA. Therefore, this has been omitted here.

**Comment by Xia-Weiwen** [comment]
@claude review

**Comment by Xia-Weiwen** [comment]
Please fix the potential lint issue by pre-commit or ruff

File: torchao/prototype/inductor/codegen/cpp_int8_sdpa_template.py [hunk 4]
```diff
@@ -1776,3 +1772,18 @@ def codegen_allocate_buffer(self, buffer_name: str, buffer_dtype, buffer_size):
                 buffer_size=buffer_size,
             )
         )
+
+    def codegen_qkv_ptr(self, kernel):
+        if len(kernel.args.input_buffers)==1:
```
Comment (Xia-Weiwen) [inline torchao/prototype/inductor/codegen/cpp_int8_sdpa_template.py:1777]: I tried the test cases on a GNR machine locally and this branch is never hit. The length of input buffers is either 3 or 4.

File: test/prototype/inductor/test_qsdpa_fusion.py [hunk 5]
```diff
@@ -400,17 +430,22 @@ def _test_fp8_sdpa_rewriter(self):
...
+        for dtype, bs, fused in itertools.product(
+            [torch.float32, torch.bfloat16], [56, 1], [True, False]
+        ):
             seqlen, numhead, headsize = 197, 16, 64
-            mod = MHAModule(
-                input_dim=headsize * numhead,
+            input_dim = headsize * numhead
+            mod = (FusedMHAModule if fused else MHAModule)(
+                input_dim=input_dim,
                 has_mask=False,
```
Comment (Xia-Weiwen) [inline test/prototype/inductor/test_qsdpa_fusion.py:440]: Why is `has_mask` always false?

File: torchao/prototype/inductor/codegen/cpp_int8_sdpa_template.py [hunk 6]
```diff
@@ -1776,3 +1772,18 @@ def codegen_allocate_buffer(self, buffer_name: str, buffer_dtype, buffer_size):
                 buffer_size=buffer_size,
             )
         )
+
+    def codegen_qkv_ptr(self, kernel):
```
Comment (Xia-Weiwen) [inline torchao/prototype/inductor/codegen/cpp_int8_sdpa_template.py:1776]: I tried the test cases on a GNR machine locally and the fp8 test cases never enters this function. Is it expected?

**Review by Xia-Weiwen — CHANGES_REQUESTED** [review]
Looks like the test cases do not cover the changes. Can you double check?

File: torchao/prototype/inductor/codegen/cpp_qsdpa_template.py [hunk 7]
```diff
@@ -1776,3 +1772,18 @@ def codegen_allocate_buffer(self, buffer_name: str, buffer_dtype, buffer_size):
                 buffer_size=buffer_size,
             )
         )
+
+    def codegen_qkv_ptr(self, kernel):
```
Comment (yuchengliu1) [reply torchao/prototype/inductor/codegen/cpp_qsdpa_template.py:2296]: Yes. Fp8 SDPA has another pass, and it has support fused qkv already. This PR is only for int8

File: torchao/prototype/inductor/fx_passes/int8_concat_linear_fusion_cpu.py [hunk 8]
```diff
@@ -0,0 +1,444 @@
...
+                            delayed_erase_nodes.append(quant_node)
+
+                    for old_node in delayed_erase_nodes:
+                        if not old_node._erased:
+                            graph.erase_node(old_node)
+
+                processed.update(qlinear_users)
+
+
+def register_int8_concat_linear_cpu_pass():
```
Comment (Valentine233) [inline torchao/prototype/inductor/fx_passes/int8_concat_linear_fusion_cpu.py:471]: Is this for int8 only, or both fp8 and int8? If it is for both, please modify all `int8` in codes to `q` (quantized).

**Review by mingfeima — APPROVED** [review]

File: test/prototype/inductor/test_qsdpa_fusion.py [hunk 9]
```diff
@@ -385,6 +385,74 @@ def _test_int8_sdpa_rewriter(self):
...
+    @unittest.skipIf(
+        not torch_version_at_least("2.7.0"),
+        reason="qsdpa requires torch 2.7 or later",
+    )
+    @unittest.skipIf(
+        "CPU" not in torch._C._dispatch_dump("torchao::qscaled_dot_product"),
+        reason="cpp kernels not built",
+    )
+    @config.patch({"freezing": True, "cpp.enable_concat_linear": True})
+    def _test_int8_fusedmha_rewriter(self):
```
Comment (Valentine233) [inline test/prototype/inductor/test_qsdpa_fusion.py]: Could we merge `_test_int8_fusedmha_rewriter` and `_test_int8_sdpa_rewriter` into one, as they share almost the same codes?

File: test/prototype/inductor/test_qsdpa_fusion.py [hunk 10]
```diff
@@ -385,6 +385,74 @@ def _test_int8_sdpa_rewriter(self):
...
+                quantizer.set_function_type_qconfig(
+                    torch.matmul, quantizer.get_global_quantization_config()
+                )
+                export_model = torch.export.export(mod, inputs, strict=True).module()
+                prepare_model = prepare_pt2e(export_model, quantizer)
+                prepare_model(*inputs)
+                convert_model = convert_pt2e(prepare_model)
+                torchao.quantization.pt2e.move_exported_model_to_eval(convert_model)
+
+                self._check_common(
```
Comment (Valentine233) [inline test/prototype/inductor/test_qsdpa_fusion.py:394]: We should check if int8 concat linear fusion is hit here.

**Review by Valentine233 — COMMENTED** [review]
Thanks!
1. Please make sure CI passes.
2. Please mark outdated comments as resolved.
3. For fp8 fusion pass, is fp8 mkl tensor unpack a following-up work in PyTorch by our side?
4. Is this verified on ViT? How are the accuracy and perf?

File: test/prototype/inductor/test_qsdpa_fusion.py (hunk 10 above)
Comment (yuchengliu1) [reply test/prototype/inductor/test_qsdpa_fusion.py:394]: added

File: test/prototype/inductor/test_qsdpa_fusion.py (hunk 9 above)
Comment (yuchengliu1) [reply test/prototype/inductor/test_qsdpa_fusion.py]: merged

File: test/prototype/inductor/test_qsdpa_fusion.py [hunk 11]
```diff
@@ -366,9 +372,14 @@ def _test_int8_sdpa_rewriter(self):
...
                 ),
-                config.patch(post_grad_custom_pre_pass=custom_pass),
+                config.patch(
+                    post_grad_custom_pre_pass=custom_pass,
+                    post_grad_custom_post_pass=None,
+                ),
             ):
                 _qsdpa_init()
+                if enable_concat_linear_fusion:
+                    register_int8_concat_linear_cpu_pass()
```
Comment (Valentine233) [inline test/prototype/inductor/test_qsdpa_fusion.py:382]: I'd like to know the process to enable `int8_concat_linear`.

I see several things: `"cpp.enable_concat_linear": True`, `post_grad_custom_post_pass=None`, `register_int8_concat_linear_cpu_pass()`. Are they all needed?

Comment (yuchengliu1) [reply test/prototype/inductor/test_qsdpa_fusion.py:382]: `enable_concat_linear` and `register_int8_concat_linear_cpu_pass` is necessary. `post_grad_custom_post_pass=None` is used to unregister the pattern for the cases without fusion.

Comment (Valentine233) [reply test/prototype/inductor/test_qsdpa_fusion.py:382]: It seems `cpp.enable_concat_linear` does not affect anything.

**Review by Valentine233 — APPROVED** [review]

**Comment by Xia-Weiwen** [comment]
@claude review

**Review by mingfeima — APPROVED** [review]
